src/Conversion/ONNXToKrnl/Quantization/DynamicQuantizeLinear.cpp

```diff
@@ -0,0 +1,149 @@
+    LocalDialectBuilder create(rewriter, loc);
+
+    ValueRange operands = adaptor.getOperands();
+    Value X = adaptor.getX();
+
+    // MemRefType for inputs and outputs.
+    auto xMemRefType = dyn_cast<MemRefType>(X.getType());
+    auto yMemRefType = dyn_cast<MemRefType>(
+        typeConverter->convertType(dqlOp.getResult(0).getType()));
+    auto yScaleMemRefType = dyn_cast<MemRefType>(
+        typeConverter->convertType(dqlOp.getResult(1).getType()));
+    auto yZeroPointMemRefType = dyn_cast<MemRefType>(
+        typeConverter->convertType(dqlOp.getResult(2).getType()));
+
+    // Types
+    Type elementType = xMemRefType.getElementType();
+    Type quantizedElementType = yMemRefType.getElementType();
+    int64_t rank = xMemRefType.getRank();
+
+    // Get shape.
+    ONNXDynamicQuantizeLinearOpShapeHelper shapeHelper(
+        op, operands, &create.krnlIE);
+    shapeHelper.computeShapeAndAssertOnFailure();
+
+    // Allocate output buffers.
+    Value Y =
+        create.mem.alignedAlloc(yMemRefType, shapeHelper.getOutputDims(0));
+    Value YScale =
+        create.mem.alignedAlloc(yScaleMemRefType, shapeHelper.getOutputDims(1));
+    Value YZeroPoint = create.mem.alignedAlloc(
+        yZeroPointMemRefType, shapeHelper.getOutputDims(2));
+
+    // Equations:
+    // y_scale = (max(x) - min(x))/(qmax - qmin)
+    // intermediate_zero_point = qmin - min(x)/y_scale
+    // y_zero_point = cast(round(saturate(itermediate_zero_point)))
+    // y = saturate (round (x / y_scale) + y_zero_point)
+    //
+    // where, saturate is to clip to [0, 255] for ui8.
+
+    // QMax, QMin.
+    Value qMax = create.math.constant(elementType, 255.0);
+    Value qMin = create.math.constant(elementType, 0.0);
+    Value QMax = create.mem.alignedAlloc(yScaleMemRefType);
+    create.krnl.store(qMax, QMax);
+    Value QMin = create.mem.alignedAlloc(yScaleMemRefType);
+    create.krnl.store(qMin, QMin);
+
+    // Compute max(x) and min (x).
+    Value none = create.onnx.none();
+    Value XMax = create.onnx.toMemref(
+        create.onnx.reduceMax(yScaleMemRefType, X, none, false));
+    Value XMin = create.onnx.toMemref(
+        create.onnx.reduceMin(yScaleMemRefType, X, none, false));
+    Value xMax = create.krnl.load(XMax);
+    Value xMin = create.krnl.load(XMin);
+    // Include 0 to max(x) and min(x).
+    // x_min = min(min(x), 0)
+    // x_max = max(max(x), 0)
+    Value zero = create.math.constant(elementType, 0.0);
+    Value greaterThanZero = create.math.sgt(xMax, zero);
+    xMax = create.math.select(greaterThanZero, xMax, zero);
+    Value lessThanZero = create.math.slt(xMin, zero);
+    xMin = create.math.select(lessThanZero, xMin, zero);
+
+    // Compute y_scale.
+    Value scale = create.math.div(
+        create.math.sub(xMax, xMin), create.math.sub(qMax, qMin));
+    create.krnl.store(scale, YScale);
+
+    // Compute y_zero_point.
+    Value interZeroPoint = create.math.div(create.math.sub(qMin, xMin), scale);
```

Comment:
@tungld Is this correct? If I'm reading this right, I think it calculates `interZeroPoint =  (qMin - xMin) / (scale)`. However it shouldn't it be `interZeroPoint = qMin - (xMin / scale)`?

Also does this correctly handle the case where the range was expanded to up/to_the_right to include zero? 

For example, say the inputs were `x = np.array([-1.0, -2.1, -1.3, -2.5, -3.34, -4.0])` For the reasons mentioned earlier in the thread, the `y_scale` is calculated so that it includes zero even though it's not actually in the range (ie `np.maximum(0,np.max(x))`). In this scenario (where `x.max() < 0`, then the zeroPoint should be qMax, so 255 for unsigned int8. 

Sadly this is another case where the [ONNX documentation](https://onnx.ai/onnx/operators/onnx__DynamicQuantizeLinear.html) shows one thing `intermediate_zero_point = qmin - min(x)/y_scale` and the [reference example](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cpu/QuantUtils.h#L149-L158) does another
```
  double zero_point_from_min = qmin - min / static_cast<double>(scale);
  double zero_point_from_max = qmax - max / static_cast<double>(scale);
  double zero_point_from_min_error =
      std::abs(qmin) - std::abs(min / static_cast<double>(scale));
  double zero_point_from_max_error =
      std::abs(qmax) - std::abs(max / static_cast<double>(scale));
  double initial_zero_point =
      zero_point_from_min_error < zero_point_from_max_error
      ? zero_point_from_min
      : zero_point_from_max;
 ````

Reply:
Good catch! as I remember I used `interZeroPoint =  (qMin - xMin) / (scale)` because I saw it in their examples: https://github.com/onnx/onnx/blob/main/docs/Operators.md#examples-45
```
Y_ZeroPoint = np.clip(round((0 - x_min) / Y_Scale), 0, 255).astype(np.uint8)
```

It happened to be true when qMin = 0.